File: src/repository/AgencyDBRepo.cs
using System.Collections.Generic;
using System.Data;
using AgentiiDeTurism.src.domain;
using log4net;

namespace AgentiiDeTurism.src.repository
{
    public class AgencyDBRepo : IAgencyRepo
    {
        private static readonly ILog log = LogManager.GetLogger("AgencyDBRepo");
        private readonly string tableName = "agencies";
        IDictionary<string, string> props;
        public AgencyDBRepo(IDictionary<string, string> props)
        {
            log.Info("Creating AgencyDBRepo");
            this.props = props;
        }
        public void clear()
        {
            log.Info("Entering clear");
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from " + tableName;
                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Deleted {0} agencies", result);
            }
            log.Info("Exiting clear");
        }

        public void delete(Agency elem)
        {
            deleteById(elem.id);
        }

        public void deleteById(int id)
        {
            log.InfoFormat("Entering deleteById with value {0}", id);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from " + tableName + " where id = @id";
                IDbDataParameter paramID = comm.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = id;
                comm.Parameters.Add(paramID);

                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Deleted {0} agencies", result);
            }
            log.Info("Exiting deleteById");
        }

        public ICollection<Agency> getAll()
        {
            log.Info("Entering getAll");
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from " + tableName;

                using (var dataR = comm.ExecuteReader())
                {
                    List<Agency> agencies = new List<Agency>();
                    while (dataR.Read())
                    {
                        int idAgency = dataR.GetInt32(0);
                        string name = dataR.GetString(1);
                        string password = dataR.GetString(2);
                        Agency agency = new Agency(idAgency, name, password);
                        agencies.Add(agency);
                    }
                    log.InfoFormat("Exiting getAll with {0} results", agencies.Count);
                    return agencies;
                }
            }
        }

        public Agency getById(int id)
        {
            log.InfoFormat("Entering getById with value {0}", id);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from " + tableName + " where id = @id";
                IDbDataParameter paramID = comm.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = id;
                comm.Parameters.Add(paramID);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idAgency = dataR.GetInt32(0);
                        string name = dataR.GetString(1);
                        string password = dataR.GetString(2);
                        Agency agency = new Agency(idAgency, name, password);
                        log.InfoFormat("Exiting getById with value {0}", agency);
                        return agency;
                    }
                }
            }
            log.InfoFormat("Exiting getById with value {0}", null);
            return null;
        }

        public void save(Agency elem)
        {
            log.InfoFormat("Entering save with value {0}", elem);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into " + tableName + " (name, password) values (@name, @password)";
                
                IDbDataParameter paramName = comm.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = elem.name;
                comm.Parameters.Add(paramName);
                
                IDbDataParameter paramPassword = comm.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = elem.password;
                comm.Parameters.Add(paramPassword);

                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Saved {0} agencies", result);
            }
            log.Info("Exiting save");
        }

        public void update(Agency elem, int id)
        {
            log.InfoFormat("Entering update with values {0}, id = {1}", elem, id);
            IDbConnection con = DBUtils.GetConnection(props);

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "update " + tableName + " set name = @name, password = @password where id = @id";

                IDbDataParameter paramName = comm.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = elem.name;
                comm.Parameters.Add(paramName);

                IDbDataParameter paramPassword = comm.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = elem.password;
                comm.Parameters.Add(paramPassword);
                
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                int result = comm.ExecuteNonQuery();
                log.InfoFormat("Updated {0} agencies", result);
            }
            log.Info("Exiting update");
        }
    }
}

File: src/main/java/repository/AgencyDBRepo.java
package repository;

import domain.Agency;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;

public class AgencyDBRepo implements AgencyRepo {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    private final String tableName = "agencies";

    public AgencyDBRepo(Properties props) {
        logger.info("Initializing AgencyDBRepo with properties: {} ", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void save(Agency elem) {
        logger.traceEntry("Saving {}", elem);
        Connection con = dbUtils.getConnection();
        String command = "insert into " + tableName + " (name, password) values (?, ?)";
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setString(1, elem.getName());
            ps.setString(2, elem.getPassword());
            int result = ps.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Agency elem) {
        deleteById(elem.getId());
    }

    @Override
    public void deleteById(Integer id) {
        logger.traceEntry("Deleting agency with id = {}", id);
        Connection con = dbUtils.getConnection();
        String command = "delete from " + tableName + " where id = ?";
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setInt(1, id);
            int result = ps.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(Agency elem, Integer id) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "update " + tableName + " set name = ?, password = ? where id = ?";
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setString(1, elem.getName());
            ps.setString(2, elem.getPassword());
            ps.setInt(3, id);
            int result = ps.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public Agency getById(Integer id) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "select * from " + tableName + " where id = ?";
        Agency agency = null;
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setInt(1, id);
            ResultSet res = ps.executeQuery();
            res.next();
            String name = res.getString("name");
            String password = res.getString("password");
            agency = new Agency(id, name, password);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
        return agency;
    }

    @Override
    public void clear() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "delete from " + tableName;
        try (PreparedStatement ps = con.prepareStatement(command)) {
            int result = ps.executeUpdate();
            logger.trace("Deleted {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
    }

    @Override
    public Collection<Agency> getAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "select * from " + tableName;
        List<Agency> agencies = new ArrayList<>();
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ResultSet res = ps.executeQuery();
            while (res.next()) {
                int id = res.getInt("id");
                String name = res.getString("name");
                String password = res.getString("password");
                agencies.add(new Agency(id, name, password));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        logger.traceExit();
        return agencies;
    }

    @Override
    public Agency getByName(String name) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        String command = "select * from " + tableName + " where name = ?";
        Agency agency = null;
        try (PreparedStatement ps = con.prepareStatement(command)) {
            ps.setString(1, name);
            ResultSet res = ps.executeQuery();
            res.next();
            int id = res.getInt("id");
            String password = res.getString("password");
            agency = new Agency(id, name, password);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB: " + e);
        }
        if (agency == null)
            throw new MyException("No agency with this name!");
        logger.traceExit();
        return agency;
    }
}
